staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -893,3 +895,209 @@ func (m *mockTxnStore) Transaction(txns ...Transaction) *TransactionError {
+		t.Fatalf("Controller failed to receive all initial setup events, got: %v", m.getHistory())
+	}
+
+	// Clear history so we only compare what happens during the RESYNC/REPLACE
+	m.clearHistory()
+
+	// Silent Drift (Events missed by controller)
+
+	// Modify: Update pod
+	pModUpdated := pMod.DeepCopy()
+	pModUpdated.Labels["ver"] = "2"
+
+	// Create pod-add
+	pAdd := &v1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod-add", Namespace: "default"}}
+
+	// Run replace
+	replaceFifo(t, fifo, []interface{}{pAdd, pKeep, pModUpdated})
+
+	err = m.waitForEventCount(ctx, 4, 5*time.Second)
+	if err != nil {
+		t.Fatalf("Controller failed to receive all events, got: %v", m.getHistory())
+	}
+
+	expectItems := []any{
+		pAdd,
+		pKeep,
+		pModUpdated,
+	}
+	items := sortItemsByKey(store.List(), DeletionHandlingMetaNamespaceKeyFunc)
+	if !apiequality.Semantic.DeepEqual(expectItems, items) {
+		t.Error(cmp.Diff(expectItems, items))
+	}
+	expectHistory := []eventRecord{
+		{Action: "add", Key: "default/pod-add"},
+		{Action: "delete", Key: "default/pod-del"},
+		{Action: "update", Key: "default/pod-keep"},
+		{Action: "update", Key: "default/pod-mod"},
+	}
+	if !apiequality.Semantic.DeepEqual(expectHistory, m.getHistory()) {
+		t.Error(cmp.Diff(expectHistory, m.getHistory()))
+	}
+}
+
+type eventRecord struct {
+	Action string
+	Key    string
+}
+
+type eventRecorder struct {
+	historyLock sync.Mutex
+	history     []eventRecord
+
+	updateCh chan bool
+}
+
+func (m *eventRecorder) OnAdd(obj interface{}, _ bool) {
+	m.record("add", obj)
+}
+func (m *eventRecorder) OnUpdate(_, obj interface{}) {
+	m.record("update", obj)
+}
+func (m *eventRecorder) OnDelete(obj interface{}) {
+	m.record("delete", obj)
+}
+
+func (m *eventRecorder) record(action string, obj interface{}) {
+	m.historyLock.Lock()
+	defer m.historyLock.Unlock()
+	key, _ := DeletionHandlingMetaNamespaceKeyFunc(obj)
+	m.history = append(m.history, eventRecord{Action: action, Key: key})
+
+	select {
+	case m.updateCh <- true:
+	default:
+	}
+}
+
+func (m *eventRecorder) clearHistory() {
+	m.historyLock.Lock()
+	defer m.historyLock.Unlock()
+	m.history = []eventRecord{}
+}
+
+func (m *eventRecorder) getHistory() []eventRecord {
+	m.historyLock.Lock()
+	historyCopy := make([]eventRecord, len(m.history))
+	copy(historyCopy, m.history)
+	m.historyLock.Unlock()
+
+	sortEvents := func(events []eventRecord) {
+		sort.Slice(events, func(i, j int) bool {
+			if events[i].Key != events[j].Key {
+				return events[i].Key < events[j].Key
+			}
+			return events[i].Action < events[j].Action
+		})
+	}
+	sortEvents(historyCopy)
+
+	return historyCopy
+}
+
+func (m *eventRecorder) waitForEventCount(ctx context.Context, count int, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	for {
+		m.historyLock.Lock()
+		currentCount := len(m.history)
+		m.historyLock.Unlock()
+
+		if currentCount >= count {
+			return nil
+		}
+
+		select {
+		case <-m.updateCh:
```

Comment:
The more "kosher" way to do that would be using `sync.Cond`. See https://github.com/kubernetes/kubernetes/blob/b4980b8d6e65602818ad27eab6c9c12398569420/staging/src/k8s.io/apiserver/pkg/storage/cacher/progress/watch_progress.go#L62-L63 for example.

Reply:
I was under the assumption that a cvar was more low level, in our case I'd like to wait on the channel but time out if the context is cancelled. Channels let us do that but a cvar would require a bit more work, right? This is a bit less efficient than a raw cvar but gives us more flexibility on the select. Not too picky either way though, especially for a test.

Reply:
Right, it's a tricky. It's hard to cooperate signal+mutex and context. Just wanted to make sure you know both options. Pick whatever seems easier to implement.